### PR TITLE
Add `client_id` to SSO package

### PR DESCRIPTION
**Pull request by loferris**
Story details: https://app.clubhouse.io/workos/story/4278

### CODE REVIEW

**Review by maxdeviant — DISMISSED** [review]

**Comment by maxdeviant** [comment]
@loferris Could you remove the `package-lock.json` since we use Yarn?

File: package-lock.json
Comment (marktran) [inline package-lock.json:1]: Can you please remove this file

File: src/sso/interfaces/authorization-url-options.interface.ts
Comment (marktran) [inline src/sso/interfaces/authorization-url-options.interface.ts:3]: Looks like the Clubhouse story is for adding `clientID` and keeping `projectID` (to maintain backward compatibility)?

**Review by marktran — CHANGES_REQUESTED** [review]

File: src/sso/interfaces/authorization-url-options.interface.ts
Comment (loferris) [reply src/sso/interfaces/authorization-url-options.interface.ts]: I added projectID back as a supported paramater and rewrote the type interfaces and the tests to reflect in my most recent commit @marktran

Comment (loferris) [reply src/sso/interfaces/authorization-url-options.interface.ts]: second commit resolved failing semaphore check

File: src/sso/interfaces/authorization-url-options.interface.ts
Comment (maxdeviant) [inline src/sso/interfaces/authorization-url-options.interface.ts:3]: We don't actually need `| undefined` here, as making a property as optional (`?`) already widens the type to include `undefined`.

File: src/sso/sso.spec.ts
Comment (marktran) [inline src/sso/sso.spec.ts:9]: I don't think we need to explicitly add "and clientID" to the `describe` blocks in the tests in this file.

`clientID` will be a required parameter and these tests are meant to enumerate on optional parameters.

Comment (loferris) [reply src/sso/sso.spec.ts]: done! I also asked this question about the tests on Slack and got similar responses: https://work-os.slack.com/archives/CPEAGQU8G/p1608669941038900

File: src/sso/interfaces/authorization-url-options.interface.ts
Comment (loferris) [reply src/sso/interfaces/authorization-url-options.interface.ts:3]: done

File: src/sso/interfaces/authorization-url-options.interface.ts
Comment (maxdeviant) [reply src/sso/interfaces/authorization-url-options.interface.ts]: > Looks like the Clubhouse story is for adding `clientID` and keeping `projectID` (to maintain backward compatibility)?

@marktran Looking at this again, I think the intention is to replace `projectID` with `clientID`. The `client_id` column on a V2 Project corresponds to the `id` column on a V1 Project.

I believe we had said that we were going to make this a breaking change and bump the major versions of the SDKs, so that when consumers update they would replace `projectID` with `clientID`.

We're already making breaking changes by removing `project_id` and replacing it with `environment_id` (see #284 and #286).

Comment (marktran) [reply src/sso/interfaces/authorization-url-options.interface.ts]: @maxdeviant I thought the`client_id` column on V2 Project was backfilled with the `id` column from a V1 Project for existing users but had a generated id in a new format for new users?

Do you know if this migration strategy is documented anywhere? I was under the impression that we could make all of this backwards compatible so current users wouldn't have to change anything.

Comment (maxdeviant) [reply src/sso/interfaces/authorization-url-options.interface.ts]: @marktran I'm not familiar with the specifics of the `client_id` for new V2 projects.

I'm trying to find where we discussed the migration strategy for the SDKs. I definitely recall us saying that it would be a major version bump, which would indicate that we were making breaking changes. I have a feeling this was something that was said verbally and perhaps wasn't written down...

Comment (marktran) [reply src/sso/interfaces/authorization-url-options.interface.ts]: We should probably get clarification from @rjvani before proceeding on these SDK changes. These stories don't have enough context.

I think there are some implications to making this a breaking change:
- Friction for current users to upgrade.
- Some current users will hold off on upgrading and we'll be forced to maintain compatibility with old versions 
- We'll need to push people to upgrade and that'll ultimately slow us down

Comment (loferris) [reply src/sso/interfaces/authorization-url-options.interface.ts]: @marktran and @maxdeviant Should I withdraw this PR and just sit on the other client_id tickets until Rohan gets back? Or does this apply to all current SDK tickets that include breaking changes?

Comment (marktran) [reply src/sso/interfaces/authorization-url-options.interface.ts]: @loferris I think we can keep this open and hold off on Clubhouse stories that have potentially breaking changes.

**Comment by loferris** [comment]
I have made changes reflecting the resolution @marktran and @rjvani clarified by Monday.

File: src/sso/sso.ts
Comment (zlwaterfield) [inline src/sso/sso.ts:60]: Should this return and throw an error here? Or not emit a warning and let the API throw an error? It feels weird to throw a warning when it should be an error.

Comment (loferris) [reply src/sso/sso.ts]: @zlwaterfield anything else we shld do before merging?

**Comment by jthodge** [comment]
@loferris Are we alright to close this PR?

**Comment by loferris** [comment]
> @loferris Are we alright to close this PR?

I actually don't know - I've been asking for feedback/approval for a while. Did this become redundant because of one of Logan's PRs?

**Comment by marktran** [comment]
> > @loferris Are we alright to close this PR?
> 
> I actually don't know - I've been asking for feedback/approval for a while. Did this become redundant because of one of Logan's PRs?

Yeah, this was covered by Logan's PR. Closing...

**Pull request closed without merging**
